### PR TITLE
bug 1562004. Provide utility to sanitize chunks

### PR DESCRIPTION
**Pull request by jcantrill**
This PR:
* Provides a utility to manually sanitize fluentd buffers for when fluentd is stuck on buffers that are corrupt and it can no longer process.

https://bugzilla.redhat.com/show_bug.cgi?id=1562004

cc @portante

### CODE REVIEW

File: fluentd/utils/sanitize_msg_chunks [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  end
+  if options.index_field.strip.empty?
+      $logger.error("The index_field can not be empty.")
+      return false
+  end
+  if options.dlq_dir != 'DROP'
+     unless options.dlq_dir.start_with?('/')
+        $logger.error("The dead letter queue must be either 'DROP' or an absolute path: #{options.dlq_dir}")
+        return false
+     end
```
Comment (richm) [inline fluentd/utils/sanitize_msg_chunks]: check for `Dir.exists?(options.dlq_dir)` here too?

File: fluentd/utils/sanitize_msg_chunks [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  clean_file = file + '.clean'
+  p = MessagePack::Packer.new(File.open(clean_file, 'w'))
+  u = MessagePack::Unpacker.new(File.new(file, 'r'))
+  u.each do |obj|
+    begin
+      time = obj[0]
+      record = obj[1]
+      target_index = record[opts.index_field]
+      target_index = target_index.downcase
+      check_record($logger, time, record) if defined?(:check_record)
```
Comment (richm) [inline fluentd/utils/sanitize_msg_chunks]: I don't think this works - `if defined?(:check_record)` will always return true

**Comment by ewolinetz** [comment]
I may be missing something regarding this utility... is this causing Fluentd to dump from its current heap or are we reading/cleaning up its DLQ?

**Comment by richm** [comment]
This is a quick&dirty utility that admins can use to clean up corrupted file buffers. It will remove corrupted records from the file buffers and dump them in a dlq, and rewrite the buffers to have only "clean" records.
Usage:
- shutdown fluentd
- run this utility - move .clean files to their original names
- startup fluentd

fluentd should now be able to flush its buffers to elasticsearch without getting stuck on the corrupted records

Yes, the next step is to change the fluent-plugin-elasticsearch to do this on the fly

File: fluentd/utils/sanitize_msg_chunks (hunk 2 above)
Comment (jcantrill) [reply fluentd/utils/sanitize_msg_chunks]: corrected.  The right way is `check_record($logger, time, record) if defined?(check_record)`

File: fluentd/utils/sanitize_msg_chunks (hunk 1 above)
Comment (jcantrill) [reply fluentd/utils/sanitize_msg_chunks]: We create it if it doesnt exist

**Comment by ewolinetz** [comment]
>Usage:
>
>    shutdown fluentd
>    run this utility - move .clean files to their original names
>    startup fluentd

ack ack
@jcantrill can you also add to the repo's README how to use this please as part of this PR?

**Comment by richm** [comment]
problem with fluentd - run.sh says fluentd not found?
/retest

**Comment by richm** [comment]
failed to deploy es?
```
openshift-logging                   30m         50m          21        logging-es-data-master-1kbcn9px-1.15265171d131f4fc                ReplicationController                                            Warning   FailedCreate                  replication-controller                         Error creating: No API token found for service account "aggregated-logging-elasticsearch", retry after the token is automatically created and added to the service account
```
/retest

**Comment by richm** [comment]
```
        "stderr": "The DeploymentConfig \"logging-es-data-master-rd09ovtn\" is invalid: \n* spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Required value: can not be empty\n* spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: \"\": name part must be non-empty\n* spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: \"\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')\n", 
```
@jcantrill @ewolinetz does this look familiar?

**Comment by jcantrill** [comment]
@richm @ewolinetz That would be because of this https://github.com/openshift/openshift-ansible/pull/7864 I think.  Logging is no longer gating for ansible because of 'flakes'.  They are run but not readily visible without digging.  We need to revisit this with @smarterclayton and @sdodson because the whole point of CI blocking or atleast making it more obvious.

**Comment by richm** [comment]
more flakes?
/retest

**Comment by richm** [comment]
still getting very strange flakes 
/retest

**Comment by richm** [comment]
the failure here is what https://github.com/openshift/origin-aggregated-logging/pull/1090 solves
/test logging
see if we can get it to pass anyway

File: fluentd/Dockerfile.centos7 [hunk 3]
```diff
@@ -58,6 +58,7 @@ RUN mkdir -p /etc/fluent/configs.d/{dynamic,user} && \
     chmod 777 /etc/fluent/configs.d/dynamic && \
     ln -s /etc/fluent/configs.d/user/fluent.conf /etc/fluent/fluent.conf
 
+COPY utils/** ${HOME}/bin
```
Comment (richm) [inline fluentd/Dockerfile.centos7]: I think this is wiping out fluentd - it seems to overwrite or remove ${HOME}/bin/fluentd
```
run.sh: line 348: exec: fluentd: not found
```

**Comment by jcantrill** [comment]
@richm maybe we no longer need this pr if we have proper handling of dlq in the elastic plugin

**Comment by richm** [comment]
> @richm maybe we no longer need this pr if we have proper handling of dlq in the elastic plugin

how far back are we planning to backport that plugin?  We have customers with this problem in 3.6.

**Comment by jcantrill** [comment]
Closing as changes to the Elasticsearch fluentd plugin will handle those issues now.

**Pull request merged**